File: frontend/src/api/modelConfig.ts
// Admin API for the model registry (Settings > Models). Every call here
// requires role=admin on the backend; regular users never hit this module —
// they consume the resulting list through api/model.ts's GET /models instead.
import { apiClient, ApiResponse } from './client'

export interface ModelCapabilities {
  supports_native_tools: boolean
  needs_guided_decoding: boolean
  max_tools: number | null
  context_window: number | null
  max_output_tokens: number | null
  supports_parallel_tool_calls: boolean
  strip_thinking_from_history: boolean
}

export type ToolProfile = 'full' | 'lean'

export interface ModelConfigEntry {
  id: string
  name: string
  provider: string
  model: string
  base_url: string | null
  is_local: boolean
  description: string | null
  enabled: boolean
  sort_order: number
  capabilities: ModelCapabilities
  capabilities_auto_detected: boolean
  tool_profile: ToolProfile
  enabled_tools: string[]
  // Never the credential itself — a masked suffix like "…wxyz", or "" if none stored.
  api_key_hint: string
  has_api_key: boolean
}

export interface CreateModelConfigInput {
  id: string
  name: string
  provider: string
  model: string
  base_url?: string | null
  api_key?: string | null
  is_local?: boolean
  description?: string | null
  enabled?: boolean
  sort_order?: number
  tool_profile?: ToolProfile
  enabled_tools?: string[]
}

export interface UpdateModelConfigInput {
  name?: string
  provider?: string
  model?: string
  base_url?: string | null
  api_key?: string | null
  is_local?: boolean
  description?: string | null
  enabled?: boolean
  sort_order?: number
  tool_profile?: ToolProfile
  enabled_tools?: string[]
}

export interface TestConnectionResult {
  ok: boolean
  error?: string | null
  latency_ms?: number | null
}

export interface ToolInfo {
  name: string
  toolkit: string
  description: string
}

export interface AvailableTools {
  tools: ToolInfo[]
  // profile name -> tool names selected by that profile with no admin
  // override — what the picker starts pre-checked as.
  profiles: Record<ToolProfile, string[]>
}

export async function listModelConfigs(): Promise<ModelConfigEntry[]> {
  const response = await apiClient.get<ApiResponse<{ models: ModelConfigEntry[] }>>('/admin/models')
  return response.data.data.models
}

export async function createModelConfig(input: CreateModelConfigInput): Promise<ModelConfigEntry> {
  const response = await apiClient.post<ApiResponse<ModelConfigEntry>>('/admin/models', input)
  return response.data.data
}

export async function updateModelConfig(id: string, input: UpdateModelConfigInput): Promise<ModelConfigEntry> {
  const response = await apiClient.patch<ApiResponse<ModelConfigEntry>>(`/admin/models/${encodeURIComponent(id)}`, input)
  return response.data.data
}

export async function deleteModelConfig(id: string): Promise<void> {
  await apiClient.delete<ApiResponse<null>>(`/admin/models/${encodeURIComponent(id)}`)
}

export async function testModelConnection(id: string): Promise<TestConnectionResult> {
  const response = await apiClient.post<ApiResponse<TestConnectionResult>>(`/admin/models/${encodeURIComponent(id)}/test`)
  return response.data.data
}

export async function listAvailableTools(): Promise<AvailableTools> {
  const response = await apiClient.get<ApiResponse<AvailableTools>>('/admin/models/tools')
  return response.data.data
}

// Presets pre-fill provider/base_url/is_local for the common providers named
// in the request — an admin picks one and only has to fill in the model name
// and (if hosted) the API key.
export interface ModelProviderPreset {
  id: string
  label: string
  provider: string
  base_url: string | null
  is_local: boolean
  modelPlaceholder: string
}

export const MODEL_PROVIDER_PRESETS: ModelProviderPreset[] = [
  { id: 'openai', label: 'OpenAI', provider: 'openai', base_url: null, is_local: false, modelPlaceholder: 'gpt-4o' },
  { id: 'google_genai', label: 'Google Gemini', provider: 'google_genai', base_url: null, is_local: false, modelPlaceholder: 'gemini-2.5-flash' },
  { id: 'openrouter', label: 'OpenRouter', provider: 'openai', base_url: 'https://openrouter.ai/api/v1', is_local: false, modelPlaceholder: 'anthropic/claude-3.5-sonnet' },
  { id: 'lmstudio', label: 'LM Studio', provider: 'openai', base_url: 'http://localhost:1234/v1', is_local: true, modelPlaceholder: 'qwen3-4b-instruct-2507' },
  { id: 'ollama', label: 'Ollama', provider: 'ollama', base_url: 'http://localhost:11434', is_local: true, modelPlaceholder: 'qwen3:4b' },
  { id: 'custom', label: 'Custom', provider: 'openai', base_url: null, is_local: false, modelPlaceholder: 'model-name' },
]

File: backend/app/interfaces/schemas/model_config.py
"""Admin API schemas for the model registry.

The credential is strictly write-only across this boundary: requests may
carry ``api_key``, responses never do — only ``api_key_hint`` (e.g. "…wxyz")
so an operator can tell which key is stored.
"""
import re
from typing import List, Optional

from pydantic import BaseModel, Field, field_validator

from app.domain.models.model_capabilities import ModelCapabilities
from app.domain.models.model_config import ModelConfig

# The registry id is embedded as a single URL path segment on every route but
# create (GET/PATCH/DELETE/test all key off it), so it must be one path
# segment's worth of characters. In particular no "/": a provider's real
# model name legitimately contains one (e.g. "google/gemma-4-e4b",
# OpenRouter's "anthropic/claude-3.5-sonnet") — that belongs in `model`, not
# `id` — but if it ends up in `id`, FastAPI reads it as *two* path segments
# and every follow-up action 404s despite creation having just succeeded.
_ID_PATTERN = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")


def _validate_registry_id(value: str) -> str:
    if not _ID_PATTERN.match(value):
        raise ValueError(
            "must contain only letters, numbers, '.', '_', '-' (no '/' or spaces) — "
            "it becomes part of a URL. Put the provider's real model name, "
            "slashes included, in the Model field instead."
        )
    return value


class ModelCapabilitiesSchema(BaseModel):
    """Capabilities as exposed to the admin UI (same shape as the domain model)."""
    supports_native_tools: bool = True
    needs_guided_decoding: bool = False
    max_tools: Optional[int] = None
    context_window: Optional[int] = None
    max_output_tokens: Optional[int] = None
    supports_parallel_tool_calls: bool = True
    strip_thinking_from_history: bool = False


class CreateModelConfigRequest(BaseModel):
    id: str = Field(min_length=1, max_length=128)
    name: str = Field(min_length=1, max_length=200)
    provider: str = Field(min_length=1, max_length=64)
    model: str = Field(min_length=1, max_length=256)
    base_url: Optional[str] = None
    api_key: Optional[str] = None          # write-only
    is_local: bool = False
    description: Optional[str] = None
    enabled: bool = True
    sort_order: int = 0
    # Omit to auto-fill from the built-in catalog — the normal path, so an
    # admin never has to know a model's tool-calling quirks.
    capabilities: Optional[ModelCapabilitiesSchema] = None
    tool_profile: str = "full"
    enabled_tools: List[str] = Field(default_factory=list)

    _validate_id = field_validator("id")(_validate_registry_id)


class UpdateModelConfigRequest(BaseModel):
    """All fields optional — a PATCH only changes what it sends.

    Notably, omitting ``api_key`` keeps the stored credential; clearing it
    requires the explicit ``clear_api_key`` flag, so an edit to an unrelated
    field can never silently drop a key.
    """
    name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    provider: Optional[str] = Field(default=None, min_length=1, max_length=64)
    model: Optional[str] = Field(default=None, min_length=1, max_length=256)
    base_url: Optional[str] = None
    api_key: Optional[str] = None
    clear_api_key: bool = False
    is_local: Optional[bool] = None
    description: Optional[str] = None
    enabled: Optional[bool] = None
    sort_order: Optional[int] = None
    capabilities: Optional[ModelCapabilitiesSchema] = None
    tool_profile: Optional[str] = None
    enabled_tools: Optional[List[str]] = None


class ModelConfigResponse(BaseModel):
    """A registered model as returned to the admin UI. Never carries a key."""
    id: str
    name: str
    provider: str
    model: str
    base_url: Optional[str] = None
    is_local: bool
    description: Optional[str] = None
    enabled: bool
    sort_order: int
    capabilities: ModelCapabilitiesSchema
    capabilities_auto_detected: bool
    tool_profile: str
    enabled_tools: List[str]
    api_key_hint: str
    has_api_key: bool

    @staticmethod
    def from_domain(config: ModelConfig) -> "ModelConfigResponse":
        return ModelConfigResponse(
            id=config.id,
            name=config.name,
            provider=config.provider,
            model=config.model,
            base_url=config.base_url,
            is_local=config.is_local,
            description=config.description,
            enabled=config.enabled,
            sort_order=config.sort_order,
            capabilities=ModelCapabilitiesSchema(**config.capabilities.model_dump()),
            capabilities_auto_detected=config.capabilities_auto_detected,
            tool_profile=config.tool_profile,
            enabled_tools=list(config.enabled_tools),
            api_key_hint=config.api_key_hint,
            has_api_key=config.has_api_key,
        )


class ListModelConfigsResponse(BaseModel):
    models: List[ModelConfigResponse]


class TestModelConnectionResponse(BaseModel):
    """Result of probing a model, so a broken credential is caught on save."""
    ok: bool
    error: Optional[str] = None
    latency_ms: Optional[int] = None


class ToolInfo(BaseModel):
    """One callable tool, for the per-model tool picker."""
    name: str
    toolkit: str
    description: str


class AvailableToolsResponse(BaseModel):
    tools: List[ToolInfo]
    profiles: dict  # profile name -> list of tool names


def capabilities_from_schema(schema: ModelCapabilitiesSchema) -> ModelCapabilities:
    return ModelCapabilities(**schema.model_dump())
